<a href="https://colab.research.google.com/github/Bayaniblues/DS-Unit-2-Linear-Models/blob/master/module2-regression-2/Copy_of_LS_DS_212_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 1, Module 2*

---

# Regression 2

## Assignment

You'll continue to **predict how much it costs to rent an apartment in NYC,** using the dataset from renthop.com.

- [ ] Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
- [ ] Engineer at least two new features. (See below for explanation & ideas.)
- [ ] Fit a linear regression model with at least two features.
- [ ] Get the model's coefficients and intercept.
- [ ] Get regression metrics RMSE, MAE, and $R^2$, for both the train and test data.
- [ ] What's the best test MAE you can get? Share your score and features used with your cohort on Slack!
- [ ] As always, commit your notebook to your fork of the GitHub repo.


#### [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)

> "Some machine learning projects succeed and some fail. What makes the difference? Easily the most important factor is the features used." — Pedro Domingos, ["A Few Useful Things to Know about Machine Learning"](https://homes.cs.washington.edu/~pedrod/papers/cacm12.pdf)

> "Coming up with features is difficult, time-consuming, requires expert knowledge. 'Applied machine learning' is basically feature engineering." — Andrew Ng, [Machine Learning and AI via Brain simulations](https://forum.stanford.edu/events/2011/2011slides/plenary/2011plenaryNg.pdf) 

> Feature engineering is the process of using domain knowledge of the data to create features that make machine learning algorithms work. 

#### Feature Ideas
- Does the apartment have a description?
- How long is the description?
- How many total perks does each apartment have?
- Are cats _or_ dogs allowed?
- Are cats _and_ dogs allowed?
- Total number of rooms (beds + baths)
- Ratio of beds to baths
- What's the neighborhood, based on address or latitude & longitude?

## Stretch Goals
- [ ] If you want more math, skim [_An Introduction to Statistical Learning_](http://faculty.marshall.usc.edu/gareth-james/ISL/ISLR%20Seventh%20Printing.pdf),  Chapter 3.1, Simple Linear Regression, & Chapter 3.2, Multiple Linear Regression
- [ ] If you want more introduction, watch [Brandon Foltz, Statistics 101: Simple Linear Regression](https://www.youtube.com/watch?v=ZkjP5RJLQF4)
(20 minutes, over 1 million views)
- [ ] Add your own stretch goal(s) !

In [1]:
%%capture
import sys

# If you're on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'
    !pip install category_encoders==2.*

# If you're working locally:
else:
    DATA_PATH = '../data/'
    
# Ignore this Numpy warning when using Plotly Express:
# FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning, module='numpy')

In [2]:
import numpy as np
import pandas as pd

# Read New York City apartment rental listing data
df = pd.read_csv(DATA_PATH+'apartments/renthop-nyc.csv')
assert df.shape == (49352, 34)

# Remove the most extreme 1% prices,
# the most extreme .1% latitudes, &
# the most extreme .1% longitudes
df = df[(df['price'] >= np.percentile(df['price'], 0.5)) & 
        (df['price'] <= np.percentile(df['price'], 99.5)) & 
        (df['latitude'] >= np.percentile(df['latitude'], 0.05)) & 
        (df['latitude'] < np.percentile(df['latitude'], 99.95)) &
        (df['longitude'] >= np.percentile(df['longitude'], 0.05)) & 
        (df['longitude'] <= np.percentile(df['longitude'], 99.95))]

In [3]:
df.head()

bathrooms  bedrooms  ... wheelchair_access common_outdoor_space
0        1.5         3  ...                 0                    0
1        1.0         2  ...                 0                    0
2        1.0         1  ...                 0                    0
3        1.0         1  ...                 0                    0
4        1.0         4  ...                 0                    0

[5 rows x 34 columns]

###Engineer at least two new features. (See below for explanation & ideas.)

In [35]:
df["perks"] =	df["elevator"]+	df["cats_allowed"]+	df["hardwood_floors"]+	df["dogs_allowed"]+	df["doorman"]+	df["dishwasher"]+	df["no_fee"]+	df["laundry_in_building"]+	df["fitness_center"]+	df["pre-war"]+	df["laundry_in_unit"]+	df["roof_deck"]+	df["outdoor_space"]+	df["dining_room"]+	df["high_speed_internet"]+	df["balcony"]+	df["swimming_pool"]+	df["new_construction"]+	df["terrace"]+	df["exclusive"]+	df["loft"]+	df["garden_patio"]+	df["wheelchair_access"]+	df["common_outdoor_space"]
df["all_rooms"] = df["bedrooms"] + df["bathrooms"]

###Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.

In [36]:
df['created'] = pd.to_datetime(df['created'])
df.shape

(48817, 36)

In [39]:
# Define important dates
a = df["created"] > '2016-04-01'
b = df['created'] <= '2016-05-01'
# features
X_train = df[(a)&(b)]
# target
y_train = X_train['price']
X_train.shape, y_train.shape

((16217, 36), (16217,))

In [41]:
# for Testing dataset
X_test = df[df['created'] >= '2016-06-01']
y_test = X_test['price']
X_test.shape, y_test.shape

((16973, 36), (16973,))

###Fit a linear regression model with at least two features.

In [49]:
# Get mean baseline
guess = y_train.mean()
print('Mean Baseline', guess)

Mean Baseline 3546.0001849910586


In [48]:
from sklearn.metrics import mean_absolute_error
y_pred = [y_train.mean()] * len(y_train)
mae = mean_absolute_error(y_train, y_pred)
print("MAE is ", mae)

MAE is  1176.0140505064794


In [52]:
y_pred = [guess] * len(y_test)
mae = mean_absolute_error(y_test, y_pred)
print("test MAE is ", mae)

test MAE is  1189.6596848403256


In [56]:
X_train

bathrooms  bedrooms  ... perks all_rooms
2            1.0         1  ...     3       2.0
3            1.0         1  ...     2       2.0
4            1.0         4  ...     1       5.0
5            2.0         4  ...     0       6.0
6            1.0         2  ...     3       3.0
...          ...       ...  ...   ...       ...
49346        1.0         1  ...     5       2.0
49348        1.0         1  ...     9       2.0
49349        1.0         1  ...     5       2.0
49350        1.0         0  ...     5       1.0
49351        1.0         2  ...     1       3.0

[16217 rows x 36 columns]

In [64]:
import plotly.express as px

px.scatter_3d(
    X_train,
    x='all_rooms', 
    y='perks', 
    z='price', 
    title='Apartment NYC'
)

In [59]:
import itertools
import numpy as np
import plotly.express as px
import plotly.graph_objs as go
from sklearn.linear_model import LinearRegression

def regression_3d(df, x, y, z, num=100, **kwargs):
    """
    Visualize linear regression in 3D: 2 features + 1 target
    
    df : Pandas DataFrame
    x : string, feature 1 column in df
    y : string, feature 2 column in df
    z : string, target column in df
    num : integer, number of quantiles for each feature
    """
    
    # Plot data
    fig = px.scatter_3d(df, x, y, z, **kwargs)
    
    # Fit Linear Regression
    features = [x, y]
    target = z
    model = LinearRegression()
    model.fit(df[features], df[target])    
    
    # Define grid of coordinates in the feature space
    xmin, xmax = df[x].min(), df[x].max()
    ymin, ymax = df[y].min(), df[y].max()
    xcoords = np.linspace(xmin, xmax, num)
    ycoords = np.linspace(ymin, ymax, num)
    coords = list(itertools.product(xcoords, ycoords))
    
    # Make predictions for the grid
    predictions = model.predict(coords)
    Z = predictions.reshape(num, num).T
    
    # Plot predictions as a 3D surface (plane)
    fig.add_trace(go.Surface(x=xcoords, y=ycoords, z=Z))
    
    return fig

In [61]:
regression_3d(
    X_train,
    x='all_rooms', 
    y='perks', 
    z='price', 
    title='Apartment NYC'
)

 ### Get the model's coefficients and intercept.


In [69]:
lr = LinearRegression()
lr.fit(X_train[["perks"]], y_train)
mae = mean_absolute_error(y_train, lr.predict(X_train[['perks']]))
print("training error", mae)

training error 1113.615247621637


In [70]:
lr.intercept_, lr.coef_

(2840.7313149059682, array([150.15353962]))

 ### Get regression metrics RMSE, MAE, and  R2 , for both the train and test data.


In [83]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

guess = y_train.mean()
y_pred = [guess] * len(y_train)

mae = mean_absolute_error(y_train, lr.predict(X_train[['perks']]))
mse = mean_squared_error(y_train, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_train, y_pred)
print("Training Metrics")
print('Mean Squared Error:', mse)
print('Root Mean Squared Error:', rmse)
print('Mean Absolute Error:', mae)
print('R^2:', r2)

Training Metrics
Mean Squared Error: 2983356.8650798206
Root Mean Squared Error: 1727.239666369384
Mean Absolute Error: 1113.615247621637
R^2: 0.0


In [87]:
guess = y_train.mean()
y_pred = [guess] * len(y_test)

mae = mean_absolute_error(y_test, lr.predict(X_test[['perks']]))
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)
print("Testing Metrics")
print('Mean Squared Error:', mse)
print('Root Mean Squared Error:', rmse)
print('Mean Absolute Error:', mae)
print('R^2:', r2)

Testing Metrics
Mean Squared Error: 3109706.7390083824
Root Mean Squared Error: 1763.4360603686152
Mean Absolute Error: 1120.5520580567734
R^2: -0.0005422971757800266


In [88]:
def all_the_metrics(in_feature):
  guess = y_train.mean()
  y_pred = [guess] * len(y_test)

  mae = mean_absolute_error(y_test, lr.predict(X_test[[in_feature]]))
  mse = mean_squared_error(y_test, y_pred)
  rmse = np.sqrt(mse)
  r2 = r2_score(y_test, y_pred)
  print("Testing Metrics")
  print('Mean Squared Error:', mse)
  print('Root Mean Squared Error:', rmse)
  print('Mean Absolute Error:', mae)
  print('R^2:', r2)

In [89]:
all_the_metrics("price")

Testing Metrics
Mean Squared Error: 3109706.7390083824
Root Mean Squared Error: 1763.4360603686152
Mean Absolute Error: 537862.6329017887
R^2: -0.0005422971757800266


In [90]:
all_the_metrics("bedrooms")

Testing Metrics
Mean Squared Error: 3109706.7390083824
Root Mean Squared Error: 1763.4360603686152
Mean Absolute Error: 1066.6014705578732
R^2: -0.0005422971757800266


In [91]:
all_the_metrics("bathrooms")

Testing Metrics
Mean Squared Error: 3109706.7390083824
Root Mean Squared Error: 1763.4360603686152
Mean Absolute Error: 1116.9325553438719
R^2: -0.0005422971757800266


 ### What's the best test MAE you can get? Share your score and features used with your cohort on Slack!


I got 537862.63 for price